[Blockifying Product Archive Templates]: Implement the blockified template conversion for the Classic Template Block.

kmanijak opened the pull request:
### Description
**The scope of this PR:**
- Implement the blockified template conversion for the Classic Template Block. That covers:
    - Product Archive
    - Product Search Results
- Update the placeholder description of the Classic Templates
- Add "align" option to the Store Notices block
- Improve how `align` attribute is passed to the PHP blocks 

**Remaining work:**
- [POSTPONED] https://github.com/woocommerce/woocommerce-blocks/issues/8287: Due to this issue, E2E tests will be submitted later as a separate PR.

Fixes https://github.com/woocommerce/woocommerce-blocks/issues/8066

### Screenshots/video

**Product Archive - side by side comparison (editor):**
| Classic Template | Blockified Template |
|------------------|---------------------|
|. <img width="1452" alt="image" src="https://user-images.githubusercontent.com/20098064/214842888-86fdf879-1a47-4c4c-882a-df28101a5da6.png"> |  <img width="1407" alt="image" src="https://user-images.githubusercontent.com/20098064/214842960-4ec491b4-4f16-4b64-a234-ecb33501d061.png"> |

**Product Archive - side by side comparison (frontend):**
| Classic Template | Blockified Template |
|------------------|---------------------|
|. <img width="1755" alt="image" src="https://user-images.githubusercontent.com/20098064/214839214-5939fdc6-282d-4fbd-8255-a37d7978f635.png"> |  <img width="1989" alt="image" src="https://user-images.githubusercontent.com/20098064/214839357-1e504761-c980-4a4e-843d-414af8c0c1de.png"> |

**Product Archive - copy update (editor)**
| Before | After |
|------------------|---------------------|
|.<img width="1071" alt="image" src="https://user-images.githubusercontent.com/20098064/214840855-2b708b27-b38d-4d45-93d6-a59986bdcad3.png">  | <img width="1106" alt="image" src="https://user-images.githubusercontent.com/20098064/214840974-1abee34c-6f6f-4474-a423-0d8d12ca5896.png"> |

**Single product - copy update (editor)**
| Before | After |
|------------------|---------------------|
|.<img width="1078" alt="image" src="https://user-images.githubusercontent.com/20098064/214840396-ea8a3a0d-2aeb-4a01-9b87-3e2459a9d406.png">  | <img width="1055" alt="image" src="https://user-images.githubusercontent.com/20098064/214840017-dceb997a-0d03-448c-8f37-23a11f4e7b6a.png"> |

Video presents the replacement of a Classic Template on two templates: Product Catalog and Products by Category

https://user-images.githubusercontent.com/20098064/214537418-51c74d39-7e00-4e26-9c2c-fe2e0f245390.mov

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests
E2E tests will be added as a separate PR

#### User Facing Testing

##### Conversion is not yet possible (hidden behind feature flag)
###### Prerequisites:
WordPress: >=6.1
Make sure Single Product template is cleared out to the default state. To achieve that:
1. Go to (`/wp-admin/site-editor.php?postType=wp_template`)
5. Click three dots next to template and click "Clear customizations"

###### Steps
1. Enter the Single Product template

###### Expected
- Check that placeholder description says "This block serves as a placeholder for your WooCommerce Single Product Block. It will display the actual product image, title, price in your store. You can move this placeholder around and add more blocks around to customize the template."
- "Upgrade to Products block" button is not visible.

Repeat for Product Catalog, Products by Category and Product Search Results templates. The placeholder description may vary slightly.

#### User Facing Testing - behind the experimental flag.
##### Conversion is not possible on WP lower than 6.1
###### Prerequisites:
WordPress: <6.1
Make sure Product Catalog and Products by Category templates are cleared out to the default state. To achieve that:
1. Go to (`/wp-admin/site-editor.php?postType=wp_template`)
7. Click three dots next to template and click "Clear customizations"

###### Steps
1. Enter the Product Catalog template

###### Expected
- Check that placeholder description says "This block serves as a placeholder for your WooCommerce Product Grid Block. It will display the actual product image, title, price in your store. You can move this placeholder around and add more blocks around to customize the template. Don't worry, you can always revert back."
- "Upgrade to Products block" button is not visible.

---

##### Convert classic templates on Product Archive templates
###### Prerequisites:
- WordPress: >=6.1
- Make sure Product Catalog/Products by Category/Products by Attribute/Products by Tag/Product Search Results templates are cleared out to the default state. To achieve that:
1. Go to (`/wp-admin/site-editor.php?postType=wp_template`)
2. Click three dots next to template and click "Clear customizations"

###### Steps
1. Enter the Product Catalog template
2. Check that placeholder description says "This block serves as a placeholder for your WooCommerce Product Grid Block. We recommend upgrading to the Products block for more features to edit your products visually. Don't worry, you can always revert back."
3. Open List View of the blocks
4. Click "Upgrade to Products block" button
<img width="902" alt="Screen Shot 2023-01-25 at 11 27 32 AM" src="https://user-images.githubusercontent.com/20098064/214540490-839d351d-c837-4db5-b991-f8293b8a924b.png">

###### Expected
WooCommerce Product Grid Block is replaced with blocks:
- Products (Beta)
    - Store Breadcrumbs (video was recorded before Breadcrumbs were added)
    - Archive Title
    - Term Description (only in _Products By *_ templates)
    - Store Notices
    - Row
        - Product Results Count
        - Catalog Sorting
    - Product Template
    - Pagination
    - No Results
<img width="377" alt="image" src="https://user-images.githubusercontent.com/20098064/220606297-868820be-7d91-4d44-b359-22e988ec9e26.png">

###### Steps - continue
5. Go to "Products (beta)" block settings
6. Make sure "Inherit query from template" is ENABLED! It should be by default.
7. Save the template
9. Go to frontend (`/shop`) and check it the blocks are displayed correctly.

###### Expected
- You should expect the view like this:
<img width="1791" alt="image" src="https://user-images.githubusercontent.com/20098064/216027859-eab87a5a-b0e9-4312-bcd0-a0e36a73d6f5.png">
- Check the Product Results Count displays proper numbers, e.g. go to the page 2 (it should say "Showing 17–17 of 17 results" if you're using "standard" products set).
- Check the Catalog Sorting sorts the products correctly, e.g. by price.


**Repeat the above for template:**
- Products by Category (to check frontend, go to `/product-category/clothing/`)
- Expected view contains Breadcrums
<img width="1698" alt="image" src="https://user-images.githubusercontent.com/20098064/216028890-a36151d9-f61d-4d28-9254-f99fe841db21.png">

---

##### Convert classic templates on Product Search Results
###### Prerequisites:
- WordPress: >=6.1
- Make sure Product Search Results template is cleared out to the default state. To achieve that:
1. Go to (`/wp-admin/site-editor.php?postType=wp_template`)
2. Click three dots next to template and click "Clear customizations"

###### Steps
1. Enter the Product Search Results template
2. Check that placeholder description says "This block serves as a placeholder for your WooCommerce Product Search Results Block. We recommend upgrading to the Products block for more features to edit your products visually. Don't worry, you can always revert back."
3. Open List View of the blocks
4. Click "Upgrade to Products block" button
<img width="902" alt="Screen Shot 2023-01-25 at 11 27 32 AM" src="https://user-images.githubusercontent.com/20098064/214540490-839d351d-c837-4db5-b991-f8293b8a924b.png">

###### Expected
WooCommerce Product Grid Block is replaced with blocks:
- Products (Beta)
    - Search Results Title
    - Store Notices
    - Row
        - Product Results Count
        - Catalog Sorting 
    - Product Template
    - Pagination
    - No Results
        - Paragraph (content: "No products were found matching your selection.")
        - Product Search
<img width="353" alt="image" src="https://user-images.githubusercontent.com/20098064/220606824-b69d0ed1-d7d0-4ed9-8d6a-33a4d2279390.png">

###### Steps - continue
5. Go to "Products (beta)" block settings
6. Make sure "Inherit query from template" is ENABLED! It should be by default.
7. Add a Product Search block to the Header template
8. Save the template
9. Go to frontend (e.g. `/shop`)
10. Use Product Search in the template
11. Enter a phrase that would cover some product names like: `t shirt` or `beanie`

###### Expected
- You should expect the product list
<img width="1689" alt="image" src="https://user-images.githubusercontent.com/20098064/216032029-c7b53c21-0f90-4eeb-8785-f91e4847ed84.png">

###### Steps - continue
12. Enter a random phrase like `blah blah` or `cars`

###### Expected
- You should expect the "No results" page with information and Product Search bar:
<img width="1688" alt="image" src="https://user-images.githubusercontent.com/20098064/216032232-787b4cd9-8a8e-4854-9983-8e31d6f5e57a.png">

---

##### Align attribute is preserved
###### Prerequisites:
WordPress: >=6.1
Make sure Product Catalog template is cleared out to the default state. To achieve that:
1. Go to (`/wp-admin/site-editor.php?postType=wp_template`)
2. Click three dots next to template and click "Clear customizations"

###### Steps
1. Enter the Product Catalog template
2. Change the `align` setting of the Classic Template block to either "None" or "Full width" 
<img width="509" alt="image" src="https://user-images.githubusercontent.com/20098064/216032869-0dc71774-7a58-4e88-87cb-aa8be28f780a.png">
3. Click "Upgrade to Products block" button

###### Expected
- Blocks keeps the same `align` setting in Editor:
    - Store Breadcrumbs
    - Archive Title
    - Store Notices
    - Row
    - Products

###### Steps - continue
4. Save the template
5. Go to frontend (`/shop`)

###### Expected
- Blocks keeps the same width (expected to be 650px wide in case of "None" and full available width in case of "Full width" option)

---

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental

### Changelog

> Enable users to migrate to blockified Product Archive templates.

kmanijak committed:
First try to render Product Arhcive template as blocks

kmanijak committed:
First iteration of the whole blockified product archive template

kmanijak committed:
Extract a blockified Product Archive Template to the templates

kmanijak committed:
Add templates to tsconfig.json so the files are resolved as part of the core code

kmanijak committed:
Add a encouragement note to use blockified version IF migration is available

Aljullu commented on assets/js/blocks/classic-template/index.tsx:
```diff
@@ -30,6 +30,7 @@ import {
 	hasTemplateSupportForClassicTemplateBlock,
 	getTemplateDetailsBySlug,
 } from './utils';
+import { getProductArchiveTemplate } from '../../../../templates/templates/blockified/archive-product';
```
We usually try to avoid relative paths like this.

Aljullu commented on templates/templates/blockified/archive-product.ts:
```diff
@@ -0,0 +1,39 @@
+} from '../../../assets/js/blocks/product-query/constants';
+import { VARIATION_NAME as productsVariationName } from '../../../assets/js/blocks/product-query/variations/product-query';
+
+const productArchiveClassName = 'woo-blocks-product-archive-template';
+const commonAttributes = {
+	className: productArchiveClassName,
+};
```
Heads-up that [we usually use the `wc-blocks-` prefix](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/contributors/contributing/coding-guidelines.md#prefixing) for class names. Having said that, do we really need to add a custom class name here?

Aljullu commented on templates/templates/blockified/archive-product.ts:
```diff
@@ -0,0 +1,39 @@
+/**
+ * Internal dependencies
+ */
+import {
+	INNER_BLOCKS_TEMPLATE as productsInnerBlocksTemplate,
+	QUERY_DEFAULT_ATTRIBUTES as productsQueryDefaultAttributes,
+} from '../../../assets/js/blocks/product-query/constants';
+import { VARIATION_NAME as productsVariationName } from '../../../assets/js/blocks/product-query/variations/product-query';
```
Same from my previous comments on relative paths. I would say that's a sign that probably `archive-product.ts` should be in `/assets/js/blocks/product-query/` instead of in `/templates/`, what do you think?

kmanijak replied:
That sound good! I was actually thinking if keeping the template in `templates/**` is a right place. I spotted there was already a `templates/templates/blockified` path with the [README](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/templates/templates/blockified/README.md) suggesting that's the place, but I tend towards keeping it in `assets/**` directory.

`/assets/js/blocks/product-query/` sounds good!

Aljullu replied:
Ups, actually I got confused in my previous review. :sweat:  I would say that given that this script is only used by the Classic Template block, it should be in the Classic Template folder (`/assets/js/blocks/classic-template/`).

In the previous message I said Product Query's folder, sorry about that! I think it makes more sense in Classic Template's folder, no? What do you think?

kmanijak replied:
No need to apologise!

> I think it makes more sense in Classic Template's folder, no? What do you think?

Yes, I agree. That's the place where's the first use case for new template. And probably the next ones, when we switch to use the [new blockified template as a default instead of Classic Template](https://github.com/woocommerce/woocommerce-blocks/issues/8067). So it makes more sense than keeping in `product-query` 😅

Aljullu commented on assets/js/blocks/classic-template/index.tsx:
```diff
@@ -79,6 +82,12 @@ const Edit = ( {
+						<br />
+						{ isBlockificationPossible &&
+							__(
+								'We encourage you to use the blockified version of this block instead. Use the button below to migrate.',
```
I think that's fine for now, but IMO we should ideally rewrite the whole text displayed in the Classic Template block now. Ie: if we allow people to replace the Classic Template block with the Product, we shouldn't have an alert saying "Do not remove this block!", it feels confusing to me.

![imatge](https://user-images.githubusercontent.com/3616980/214031823-689c4c16-6dcf-4f6a-b8c7-10d4f60cbfbd.png)

Aljullu reviewed (commented):
Looks great. Very nice to see the progress of this project. :clap: 

I think we should wrap the Catalog Sorting and Product Result Count blocks in a Row block (which is a variation of the Group block), so they are in the same line.

![imatge](https://user-images.githubusercontent.com/3616980/214034097-45dd7ba6-a63a-47a8-a5bc-83eccb50a3fc.png)


Besides that, I left some comments inline.

Aljullu commented:
To answer your questions:

> - should the new template (JS) be stored in `templates/templates/blockified/archive-product.ts`? Maybe it should lay down somewhere in the `assets/**` along with other JS code?

I left an inline comment regarding that. But considering this script is only called from the Classic Template block, I would move it there.

> - should the template export just the block structures or the block instance like it is right now (using `createBlock` method from `@wordpress/blocks`). Maybe there's another idea you may suggest?

I don't have any alternative idea. Let's see if somebody has one. Otherwise, this approach LGTM.

> - blocks used in the template require additional styles to layout them properly. Is that ok if they sit next to the template, so `templates/templates/blockified/archive-product.scss`? It doesn't sound right, especially that it would require adjusting the Webpack config to include the styles from there as well. I'll explore other possibilities, but I'm open for suggestions.

Which additional styles? If they need specific CSS, that should be probably be fixed in the block leve. :thinking:

kmanijak commented:
@Aljullu, thank you so much for a great review! 🙏 

> But considering this script is only called from the Classic Template block, I would move it there.

Can I confirm, by "move it there", do you mean next to Classic Template or `assets` directory in general (for example `/assets/js/blocks/product-query/` as you suggested in inline comment)?

> Which additional styles? If they need specific CSS, that should be probably be fixed in the block leve. 🤔

I meant the styles related to lay outing the blocks, so putting Product Results Count and Catalog Ordering in a row. Exactly what you showed in a comment [suggesting to use Row block](https://github.com/woocommerce/woocommerce-blocks/pull/8248#pullrequestreview-1265433150). So with your recommendation, the additional styles won't be necessary anymore and so the custom class. 👍

kmanijak committed:
Add a Row block wrapping Product Results Count and Catalog Sorting blocks

kmanijak committed:
Move blockified product-archive from templates/ to assets/ directory

kmanijak committed:
Remove unnecessary margin from Product Results Count block

When used in a Row block in a blockified Archive Product template, Product Results Count had additional unnecessary margin which caused misalignment with the Catalog Sorting block

kmanijak committed:
Fix TypeScript error

Aljullu commented:
> Can I confirm, by "move it there", do you mean next to Classic Template or `assets` directory in general (for example `/assets/js/blocks/product-query/` as you suggested in inline comment)?

I answered inline above. :slightly_smiling_face: 

> I meant the styles related to lay outing the blocks, so putting Product Results Count and Catalog Ordering in a row. Exactly what you showed in a comment [suggesting to use Row block](https://github.com/woocommerce/woocommerce-blocks/pull/8248#pullrequestreview-1265433150). So with your recommendation, the additional styles won't be necessary anymore and so the custom class. 👍 

Great!

kmanijak committed:
Update the description of the Classic Template Editor placeholder

kmanijak committed:
Remove unnecessary entry in tsconfig.json to include templates directory

It was added couple of commits earlier, since the template was kept there, but it was decided to move it to assets directory, so entry is no longer necessary

kmanijak committed:
Move the  to

kmanijak committed:
Differentiate the Classic Template placeholder description depending on the availability to convert to Products block

kmanijak committed:
Add E2E test that the Classic Template can be replaced with the blockified version

kmanijak committed:
Abandon writing E2E tests for now, as there's a problem to run them locally

kmanijak committed:
Merge branch 'trunk' into add/convert-classic-template

Aljullu commented on assets/js/blocks/classic-template/index.tsx:
```diff
@@ -60,6 +81,13 @@ const Edit = ( {
 
+	// Blockification is possible for the WP version 6.1 and above,
+	// which are the versions the Products block supports.
+	const isBlockificationPossible = isWpVersion( '6.1', '>=' );
```
For now, `isBlockificationPossible` should be false when in the Single Product template:

![imatge](https://user-images.githubusercontent.com/3616980/214588943-886888b9-3ac2-45a7-a7c3-dcaaaed21ca8.png)

It should also have a different description and eventually a different button label.

kmanijak replied:
Great catch!

1. I extended the testing steps to cover the Single Product template (expected: description which doesn't indicate the update and no button)
2. I made the blockification more generic so it's prepared for Single Product template blockification. For the Single Product template it will be a matter of changing providing proper config [in here](https://github.com/woocommerce/woocommerce-blocks/pull/8248/files#diff-49cf8439043f78ee9f68b6374ef821be0b59c5804fec85e9aecf528d7fe9bcecR50) in the same manner as the product archive. If that change is ok and will be merged, I'll update the issue regarding Single Product template to let the developers know.

Aljullu commented on assets/js/blocks/product-results-count/style.scss:
```diff
@@ -5,5 +5,6 @@
 	.woocommerce-result-count {
 		float: none;
 		font-size: inherit;
+		margin: 0;
```
For me, `.woocommerce .woocommerce-ordering` also has some bottom margin, should we remove it from there too so both blocks are aligned?

![imatge](https://user-images.githubusercontent.com/3616980/214588682-293adf6c-fb83-4199-a5ef-5c871e0e037d.png)

Aljullu commented on assets/js/blocks/classic-template/archive-product.ts:
```diff
@@ -0,0 +1,59 @@
+		( { name }: { name: string } ) => name === rowVariationName
+	);
+
+	if ( ! rowVariation ) return null;
```
Nit: we tend to always use multiline clauses:

```suggestion
	if ( ! rowVariation ) {
		return null;
	}
```

Aljullu reviewed (commented):
Really good job here, this is shaping up nicely. :clap: I left a couple of comments, but after that I think this will be ready to be merged.

kmanijak committed:
Syntax improvement

kmanijak committed:
Set margin for Catalog Sorting to 0, so it aligns properly when used in blockified Archive Product template

kmanijak committed:
Make the blockification config, so it covers the Product Archive as well as Single Product templates

kmanijak committed:
Merge branch 'add/convert-classic-template' of github.com:woocommerce/woocommerce-blocks into add/convert-classic-template

kmanijak committed:
Move the product-archive specific functions from classic-template/index.tsx to the product-archive.ts

dinhtungdu reviewed (commented):
I notice the layout is different between the default template with the Classic Template block and the new blockified template. The classic one uses wide width while the new one uses content width.

It'd be awesome if we can take the current width/alignment of the Classic Template block and apply it to the converted template. For example, if the align setting of Classic Template is `Wide`, the converted template should be wide too.

We may need to add the align support back to the `Store Notices` block, we only need `None`, `Wide width`, and `Full width` options, the same as the Classic Template block.

kmanijak commented:
> We may need to add the align support back to the Store Notices block, we only need None, Wide width, and Full width options, the same as the Classic Template block.

Nice catch @dinhtungdu ! That becomes a requirement for the [Breadcrumbs block as well](https://github.com/woocommerce/woocommerce-blocks/pull/8222).

I'm adding the setting to Store Notices block and will write a comment in the Breadcrumbs PR.

kmanijak committed:
Add alignment option to the Store Notices block and make the blockified template blocks aligned wide

That is required, so the Classic Template layout is preserved

kmanijak commented:
@dinhtungdu, I added the support for the `align` attribute to Store Notices and also provided a correct initial settings to other blocks ([commit](https://github.com/woocommerce/woocommerce-blocks/pull/8248/commits/88ddb2abf058679a118c20d6d2a2658110e8ce56)).

Thanks to that we keep the same width and layout as on the screenshots below (I added them to the PR description as well):

**Product Archive - side by side comparison (editor):**
| Classic Template | Blockified Template |
|------------------|---------------------|
|. <img width="1452" alt="image" src="https://user-images.githubusercontent.com/20098064/214842888-86fdf879-1a47-4c4c-882a-df28101a5da6.png"> |  <img width="1407" alt="image" src="https://user-images.githubusercontent.com/20098064/214842960-4ec491b4-4f16-4b64-a234-ecb33501d061.png"> |

**Product Archive - side by side comparison (frontend):**
| Classic Template | Blockified Template |
|------------------|---------------------|
|. <img width="1755" alt="image" src="https://user-images.githubusercontent.com/20098064/214839214-5939fdc6-282d-4fbd-8255-a37d7978f635.png"> |  <img width="1989" alt="image" src="https://user-images.githubusercontent.com/20098064/214839357-1e504761-c980-4a4e-843d-414af8c0c1de.png"> |

I suggest waiting for the Breadcrumbs block to be merged before we do the final review/testing of this one.

dinhtungdu commented on assets/js/blocks/classic-template/archive-product.ts:
```diff
@@ -0,0 +1,107 @@
+			...attributes.layout,
+			justifyContent: 'space-between',
+		},
+		align: 'wide',
```
We're hardcoding the value of align here to `wide`. What if the merchant changes the alignment of the Classic Template block to `None` or `Full width`? Can we preserve that attribute?

kmanijak replied:
Great point! `align` attribute is now inherited from the placeholder block ([commit](https://github.com/woocommerce/woocommerce-blocks/pull/8248/commits/36f740b18c8ecebd6e14d7694ea5ad38e35b2e43) - but note that this evolved a bit in next commits).

dinhtungdu commented on assets/js/blocks/classic-template/archive-product.ts:
```diff
@@ -0,0 +1,107 @@
+	);
+};
+
+const getDescriptionAllowingConversion = ( templateTitle: string ) =>
+	sprintf(
+		/* translators: %s is the template title */
+		__(
+			"This block serves as a placeholder for your %s. We recommend upgrading to the Products block for more features to edit your products visually. Don't worry, you can always revert back.",
+			'woo-gutenberg-products-block'
+		),
+		templateTitle
+	);
+
+const getDescriptionDisallowingConversion = ( templateTitle: string ) =>
+	sprintf(
+		/* translators: %s is the template title */
+		__(
+			'This block serves as a placeholder for your %s. It will display the actual product image, title, price in your store. You can move this placeholder around and add more blocks around to customize the template.',
+			'woo-gutenberg-products-block'
+		),
+		templateTitle
+	);
```
What about merging these two functions into one? like `getDescription` accepting two arguments `templateTitle` and `templatePlaceholder`. By doing so, we don't have to expose the logic in the `index.tsx` file.

kmanijak replied:
Nice one. It simplified the interface of blockified templates ([commit](https://github.com/woocommerce/woocommerce-blocks/pull/8248/commits/c2c5657f9014d8b3d458e56ce26278a4cbd5d933)). However, I chose the function to accept `canConvert` rather than `templatePlaceholder`. Information if the conversion is possible [is required in `classic-template` anyway](https://github.com/woocommerce/woocommerce-blocks/pull/8248/files#diff-49cf8439043f78ee9f68b6374ef821be0b59c5804fec85e9aecf528d7fe9bcecR85), so we can keep the `getDescription` pure. I hope you're ok with that.

dinhtungdu replied:
I totally agree 💯

kmanijak committed:
Merge branch 'trunk' into add/convert-classic-template

dinhtungdu commented on assets/js/blocks/classic-template/archive-product.ts:
```diff
@@ -0,0 +1,107 @@
+		createProductsBlock(),
+	].filter( Boolean ) as BlockInstance[];
+
+const isBlockificationPossible = ( templatePlaceholder: string ) => {
+	// At the moment blockification is available for product archive only.
+	// Blockification is possible for the WP version 6.1 and above,
+	// which are the versions the Products block supports.
+	return (
+		templatePlaceholder === PLACEHOLDERS.archiveProduct &&
+		isWpVersion( '6.1', '>=' )
+	);
+};
```
This method is executed only when the `templatePlaceholder` is `archive-product`, so it doesn't need to pass the `templatePlaceholder` here.

Also, here we're introducing a new term: `blockification`. What do you think about using existing terms to express the same meaning, something like `isBlockTemplateAvailable`, or even `canConvert`?

kmanijak replied:
> This method is executed only when the templatePlaceholder is archive-product, so it doesn't need to pass the templatePlaceholder here.

You're right, it was a leftover from previous approach. I changed the naming as well. Both changes addressed in [commit](https://github.com/woocommerce/woocommerce-blocks/pull/8248/commits/4a4b8e57c32222bd0631a8c592295918b05a1f42).

kmanijak committed:
Create single-product.ts file which is a placeholder for the blockified Single Product template

kmanijak committed:
Make Blockified Product Archive template inherit the align attibute

kmanijak committed:
Simplify the interface of blockified templates

Expose  function instead of two functions for allowing and disallowing conversion

kmanijak committed:
Add a BlockifiedTemplate type

kmanijak committed:
Rename and simplify the function checking if conversion of classic template to block version is possible

kmanijak committed:
Merge branch 'add/convert-classic-template' of github.com:woocommerce/woocommerce-blocks into add/convert-classic-template

kmanijak committed:
Align the variable naming to use  instead of

kmanijak commented:
@dinhtungdu, we've discussed the possibility to get rid of [`fallback` flow in classic-template](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/blocks/classic-template/index.tsx#L52).

The template is chosen based on a fuzzy logic in [`getTemplateDetailsBySlug` util function](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/blocks/classic-template/utils.ts#L13), which covers the possibility of not finding a matching template. In that case the `fallback` mentioned before is in use. The util function was introduced in order to fix the issue: [PR](https://github.com/woocommerce/woocommerce-blocks/pull/7033) and increases a chance of finding a proper template based on slug, but doesn't guarantee that.

Having said that, I'd keep the `fallback` flow. What's your opinion?

dinhtungdu commented on assets/js/blocks/classic-template/index.tsx:
```diff
@@ -68,48 +93,26 @@ const Edit = ( {
+									const classicTemplateAttributes =
+										select(
+											`core/block-editor`
+										).getBlockAttributes( clientId ) || {};
```
We can use `attributes` prop here instead of selecting it from the store.

dinhtungdu commented on assets/js/blocks/store-notices/block.json:
```diff
@@ -6,7 +6,14 @@
 	"supports": {
-		"multiple": false
+		"multiple": false,
+		"align": [ "wide", "full" ]
```
We need to update the `render` method of `./src/BlockTypes/StoreNotices.php`  to make the alignment work on the front end.

kmanijak replied:
I didn't realise earlier there's a separate `get_align_class_and_style` method for that. I applied the change to:
- Store Notices
- Breadcrumbs

It works fine when the Classic Template is replaced by blocks. They preserve the alignment. But I found a bug that when the alignments are changed it's starts to become unpredictable. I was trying to find a pattern when and what happens, but I don't see one. So I'm going to create an issue and describe the most I can.

kmanijak replied:
I created an issue: https://github.com/woocommerce/woocommerce-blocks/issues/8336. I believe it's not a blocker and also confirmed it's not related to the newly implemented blocks, but it seems more global.

dinhtungdu replied:
Nice catch!

dinhtungdu reviewed (commented):
Code changes look solid to me. I left some minor comments, otherwise, this is LGTM. Also, as we merged the Breadcrumbs block, let's include it in this PR.

kmanijak committed:
Pass the Classic Template attributes to the blockified template instead of getting it from data store

kmanijak committed:
Merge branch 'trunk' into add/convert-classic-template

kmanijak committed:
Include Breadcrumbs block in the Blockified Product Archive Template

kmanijak committed:
Consume alignment attributes of Catalog Sorting in the PHP render function

kmanijak committed:
Consume alignment attributes of Breadcrumbs in the PHP render function

kmanijak committed:
Remove align support from Catalog Sorting and add to Store Notices block

dinhtungdu commented on src/BlockTypes/Breadcrumbs.php:
```diff
@@ -34,12 +34,15 @@ protected function render( $attributes, $content, $block ) {
-		$classname          = $attributes['className'] ?? '';
-		$classes_and_styles = StyleAttributesUtils::get_classes_and_styles_by_attributes( $attributes );
+		$classname             = $attributes['className'] ?? '';
+		$classes_and_styles    = StyleAttributesUtils::get_classes_and_styles_by_attributes( $attributes );
```
I think instead of using `get_align_class_and_style`, we should add align support to `StyleAttributesUtils::get_classes_and_styles_by_attributes()` instead. `$attributes['className']` should also be included but let's work on it in a separate refactoring PR.

dinhtungdu replied:
`align` and `text_align` are missing from `StyleAttributesUtils::get_classes_and_styles_by_attributes()`.

kmanijak replied:
> I think instead of using get_align_class_and_style, we should add align support to StyleAttributesUtils::get_classes_and_styles_by_attributes() instead
>
> align and text_align are missing from StyleAttributesUtils::get_classes_and_styles_by_attributes().

I like the idea. It's covered in [this commit](https://github.com/woocommerce/woocommerce-blocks/pull/8248/commits/cd71a6afa8c42f6c5a55eaa4e10f442174cec23a).

> $attributes['className'] should also be included but let's work on it in a separate refactoring PR.

I agree that deserves a separate PR, since it would be more impactful change (requires adjustments in nearly every block and it would bloat this PR). I'll create an issue 👍

dinhtungdu commented on assets/js/blocks/classic-template/archive-product.ts:
```diff
@@ -0,0 +1,117 @@
+const isConversionPossible = () => {
+	// Blockification is possible for the WP version 6.1 and above,
+	// which are the versions the Products block supports.
+	return isWpVersion( '6.1', '>=' );
```
Should we support Gutenberg here as well? Eg: site use WP 6.0 but latest version of Gutenberg. It's fine to keep it simple as it is right now, I just want to discuss the possible usecases.

kmanijak replied:
I think what determines the possibility to convert is the availability of _Products_ block. It doesn't require Gutenberg, but it [does require WP 6.1. or higher](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/blocks/product-query/variations/product-query.tsx#L22). So I think the condition should be aligned - that guarantees all the blocks required to convert the template are available.

Let me know what you think about that 👍

dinhtungdu replied:
Oh I missed that. Totally agree! Good catch!

dinhtungdu reviewed (commented):
Thanks for bearing with me! And sorry for incremental reviews 😬.

I notice some more issues that I think we should fix within this PR:
- The archive title is missing from the converted template.
- The Product Search Results template has some different elements comparing to general archive product template:
  - The title
  - The no results found content (which should include an additional search form).

Because of that, Product Search Results template should have a dedicate config, or at least, a different blockified template.

What do you think @kmanijak?

kmanijak committed:
Extend the get_classes_and_styles_by_attributes method with align and text_align attributes

kmanijak committed:
Add Archive Title block to the Blockified Classic Template

kmanijak committed:
Minor getRowBlock function refactor

kmanijak committed:
Add  property to the classic templates and base the config on it instead of placeholder

kmanijak committed:
Add separate blockified template for a Product Search Results

kmanijak committed:
Pass  attribute from classic template to No Results block in Product Search Results

kmanijak committed:
Extract the common functions between blockified archive-template and product-search-results to utils

kmanijak committed:
Enable 'Inherit query from template' in Products block by default when converting the Classic Template to blockified one

kmanijak committed:
Merge branch 'trunk' into add/convert-classic-template

kmanijak commented:
@dinhtungdu,

> Thanks for bearing with me! And sorry for incremental reviews 😬.

No need to apologise, I appreciate deep dive into what's happening in the PR! 🖐️ 

> The archive title is missing from the converted template.

Good one, it's been added to both: Product Archive and Product Search Results pages ([function](https://github.com/woocommerce/woocommerce-blocks/pull/8248/files#diff-b407ec38238585241ebf01c6234425454870896d954d9ca696e7e56afbb21fe3R56))

> The Product Search Results template has some different elements comparing to general archive product template:
> - The title
> - The no results found content (which should include an additional search form).
>
> Because of that, Product Search Results template should have a dedicate config, or at least, a different blockified template.
> What do you think @kmanijak?

Agree! I've added a dedicated [config file](https://github.com/woocommerce/woocommerce-blocks/pull/8248/files#diff-526f69522d1f26344e2fef49a7585cab92b7fff25a114959cd9a4016156a1d3e). Template is adjusted.
I wanted to double check we want the Product Search block included in No Results - I don't see that behaviour with Classic Template, but I followed your comment and the [template from your PR](https://github.com/woocommerce/woocommerce-blocks/pull/8308/files#diff-7b43cd294e750d01f3be587296b662ce9170dfa28c3b7a15ab5753218c420be0R33).

Besides that, the testing steps were improved and extended with more cases.

kmanijak commented:
Remaining question is if the blocks like Breadcrumbs, Notices etc. should be an inner blocks of Products or not as per @dinhtungdu's [PR](https://github.com/woocommerce/woocommerce-blocks/pull/8308).

dinhtungdu commented on assets/js/blocks/classic-template/types.ts:
```diff
@@ -1 +1,19 @@
+	align?: string;
+};
+
+export type BlockifiedTemplate = {
```
```suggestion
export type BlockifiedTemplateConfig = {
```

I think a more explicit name here will help other folks understand this easier.

dinhtungdu commented on assets/js/blocks/classic-template/index.tsx:
```diff
@@ -68,48 +96,20 @@ const Edit = ( {
 				</div>
 				<div className="wp-block-woocommerce-classic-template__placeholder-wireframe">
-					{ isExperimentalBuild() && (
+					{ canConvert && (
```
Let's wait for the final decision on the templates before removing the feature gating because if this PR is getting into Core, the conversion will work there. We can either wait for #8308 or add the feature gating back to merge this PR now (and remove the flag later in #8308).

kmanijak replied:
I think I'm ok with waiting, so both PRs are aligned in terms of a block structure 👍

kmanijak replied:
@Aljullu, can I get your opinion about this?

I think I'll bring back the feature flag for now:
- the auto-conversion is not enabled yet in #8308 
- compatibility layer has been disabled for now as it caused regression https://github.com/woocommerce/woocommerce-blocks/pull/8507

Actually seems like there's no space for opinions 😅 but it has to be hidden behind the feature flag yet and all three parts have to be enabled together.

Aljullu replied:
Good point. Yes, it makes sense to me. :+1:

kmanijak replied:
Added [in the commit](https://github.com/woocommerce/woocommerce-blocks/pull/8248/commits/462fa4a4058903e8bb2498f8949a97c0e2b7d1f4)

dinhtungdu reviewed (commented):
This is LGTM! Thanks so much for your updates! There is a comment that affects the merge ability of this PR (so I haven't approved this yet)

kmanijak committed:
Improve the naming of BlockifiedTemplateConfig type

kmanijak committed:
Differentiate Product Catalog and Products By * templates. The latter include Term Description block

kmanijak committed:
Merge branch 'add/convert-classic-template' of github.com:woocommerce/woocommerce-blocks into add/convert-classic-template

kmanijak commented:
As per the [comment in similar PR](https://github.com/woocommerce/woocommerce-blocks/pull/8308#discussion_r1094409565), this PR is now extended with blockified config variation for _Products By *_ templates. The difference between _Product Catalog_ config is the _Term Description_ block is included ([commit](https://github.com/woocommerce/woocommerce-blocks/pull/8248/commits/11a90df36e02a353567aca6382c74508c31f1025)).

dinhtungdu commented on assets/js/blocks/classic-template/constants.ts:
```diff
@@ -11,7 +11,8 @@ import { TemplateDetails } from './types';
 	singleProduct: 'single-product',
-	archiveProduct: 'archive-product',
+	productCatalog: 'product-catalog',
+	productsBy: 'products-by',
```
To me, it's a bit unclear what `productsBy` means. Can we consider other alternative here? Somethings like `productTaxonomy`, `productLoop`, `productGrid`? or even keep `archiveProduct` for shared template, and use `productCatalog` just for the catalog page.

kmanijak replied:
> To me, it's a bit unclear what productsBy means.

Thanks for feedback, makes sense! I Changed that to `productTaxonomy` ([commit](https://github.com/woocommerce/woocommerce-blocks/pull/8248/commits/7f6a43830d4a0a263ae4f69a66ff79c81dfcd97f))

kmanijak committed:
Change unclear ProductsBy to ProductTaxonomy in regards to classic template conversion

kmanijak committed:
Revert the margin fix which was added already on trunk

kmanijak committed:
Merge branch 'trunk' into add/convert-classic-template

kmanijak committed:
Move the surrounding blocks as inner blocks of Products

kmanijak commented:
This PR has been waiting for a final decision if the blocks: Breadcrumbs, Store Notices, Product Result Count, Catalog Sorting should surround Products block or be inner blocks:

As per the PR: https://github.com/woocommerce/woocommerce-blocks/pull/8308, these blocks are the inner blocks of Products. [This commit](https://github.com/woocommerce/woocommerce-blocks/pull/8248/commits/f925c7bdac72c29c4835fa69e71f778233ff7473) applies the change here. Also, the testing steps are amended now.

Given that, this PR is ready for final testing and review.

cc: @Aljullu, @albarin

kmanijak committed:
Merge branch 'trunk' into add/convert-classic-template

Aljullu reviewed (commented):
This is looking awesome, good job! The code looks good and it's testing as expected. I have only seen one issue: when converting the Classic Template block to the new one, I can see several visual differences.

Classic template | Blockified template
--- | ---
![imatge](https://user-images.githubusercontent.com/3616980/220661059-280e5bc5-c8fe-4058-92ff-ce896fff0283.png) | ![imatge](https://user-images.githubusercontent.com/3616980/220661174-e01e6bcd-6354-4f0a-bd72-d6e45d4c62fd.png)

Some things I see:

* We should include the Product Rating block, which is not included in the default pattern of the Products block.
* We should make the Product Title a link (I don't know why the Products block don't have it enabled by default, it might be better to fix it there?).
* The Product Price should be centered (IMO, this might be worth doing in the Products block directly).

I noticed your screenshots don't have some these issues. So maybe I have been doing something wrong? :confused:

kmanijak committed:
Hide the conversion behind the experimental build flag (as it was before)

kmanijak commented:
> We should include the Product Rating block, which is not included in the default pattern of the Products block.

I started implementing that... But I think it's worth implementing it in Products block directly and include Product Rating by default in [the Products inner blocks template](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/blocks/product-query/constants.ts#L75). The reason being:
- I don't want to hardcode the position where to inject the Product Rating (pseudocode: `innerBlockTemplate[0][1] = createBlock('core/product-rating')`)
- I would like to avoid writing a generic function (so it's future proof) that would find a `post-template` and `product-title` to inject rating afterward. I think it's unnecessary complex comparing to the gain.
I would raise the question about extending Products with Product Rating. What do you think? 🤔 

> We should make the Product Title a link (I don't know why the Products block don't have it enabled by default, it might be better to fix it there?).

Good catch. I see `post-title` has an attribute `isLink`. But I'll prepare a separate PR directly in Products block, so it can be confirmed by others.

> The Product Price should be centered (IMO, this might be worth doing in the Products block directly).
I noticed your screenshots don't have some these issues. So maybe I have been doing something wrong? 😕

That's weird, it's centered by default for me and it's defined [in here](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/blocks/product-query/constants.ts#L94). So I'm not sure how to approach this one 🤔

kmanijak commented:
> > The Product Price should be centered (IMO, this might be worth doing in the Products block directly).
I noticed your screenshots don't have some these issues. So maybe I have been doing something wrong? 😕

> That's weird, it's centered by default for me and it's defined [in here](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/blocks/product-query/constants.ts#L94). So I'm not sure how to approach this one 🤔

@Aljullu, ok, I can reproduce it (here and on `trunk`). It shows centered in the Editor, but it is left-aligned in the FE. Seems the alignment is broken again :( I'll create an issue, I don't think the fix should be included in this PR. 🙏 

EDIT: issue created: https://github.com/woocommerce/woocommerce-blocks/issues/8521

kmanijak commented:
Summary:

> We should include the Product Rating block, which is not included in the default pattern of the Products block.

It's raised for internal discussion to confirm if the change could be included in the Products block directly. If not I'll add it to the templates within this PR.
Although, I tend to think we could merge this in a current shape (without Product Rating) and apply the change once there's a decision. The reasons would be:
- this is still hidden behind feature flag
- there's no Product Rating in #8308, which is already in `trunk`, so it would be  good to apply the change in both places at once. Let me know what you think @Aljullu 🙌 

> We should make the Product Title a link (I don't know why the Products block doesn't have it enabled by default, it might be better to fix it there?). 

It's fixed and merged to `trunk` ([PR](https://github.com/woocommerce/woocommerce-blocks/pull/8519))

> The Product Price should be centered (IMO, this might be worth doing in the Products block directly).

Reproducible on `trunk`, I created an issue: https://github.com/woocommerce/woocommerce-blocks/issues/8521

Aljullu reviewed (approved):
>  I would raise the question about extending Products with Product Rating. What do you think? :thinking: 

It makes sense to me. :ok_hand: 

> Good catch. I see `post-title` has an attribute `isLink`. But I'll prepare a separate PR directly in Products block, so it can be confirmed by others.

Thanks!

> Although, I tend to think we could merge this in a current shape (without Product Rating) and apply the change once there's a decision.

Sounds good. I agree that we can merge this PR now and work on the remaining tasks in follow-up PRs.

Thanks, @kmanijak!

kmanijak committed:
Merge branch 'trunk' into add/convert-classic-template

kmanijak committed:
Merge branch 'trunk' into add/convert-classic-template

kmanijak committed:
Merge branch 'trunk' into add/convert-classic-template
